Prototyping React

vboctor opened the pull request.

Commit:
initial commit

Commit:
bundle to js/mantis

Commit:
added mantis.* to gitignore

Commit:
js minify script

Commit:
updated package json and hard coded params

Commit:
removed pack script

Commit:
IssueService class

Commit:
Issue models

Commit:
Relationship model

vboctor committed:
Implement LocalizedStringsGetCommand

Signed-off-by: Victor Boctor <victor@mantishub.net>

Commit:
removed console log

Commit:
removed unneeded functions and models

vboctor committed:
Implemented ConfigsGetCommand

Signed-off-by: Victor Boctor <victor@mantishub.net>

Commit:
ConfigService

vboctor committed:
Include mantis.js in all pages

Signed-off-by: Victor Boctor <victor@mantishub.net>

vboctor committed:
Remove RelationshipsParse() dead code

Signed-off-by: Victor Boctor <victor@mantishub.net>

vboctor committed:
Remove IssueRelationshipAddRequest

It is much simpler to have the two arguments directly on the function.

Signed-off-by: Victor Boctor <victor@mantishub.net>

vboctor committed:
Remove ConfigService and use PHP commands instead

- Removed the ConfigService from JS
- Added calls to LocalizedStringsGetCommand() as sample and embedded the result html for JS.
- Added calls to ConfigsGetCommand() as samlpe and embedded the result in html for JS.

Signed-off-by: Victor Boctor <victor@mantishub.net>

vboctor commented on client/js/models/Relationship.ts:
```diff
@@ -0,0 +1,13 @@
+export interface Relationship {
+  id: number;
+  type: RelationshipType;
+  issue: Issue;
```
@waltergar The relationship should use the proper `Issue`, but it can use `IssueRelationshipRef` or equivalent.  This way the proper `Issue` model could enforce the right semantics, e.g. fields like `description` and `project` are mandatory.  It also is strange to have an issue that contains a relationship that contains an issue.

vboctor commented on client/js/components/IssueRelationships.tsx:
```diff
@@ -0,0 +1,144 @@
+            </select>
+            &nbsp;
+            <input
+              type='text'
```
@waltergar Typing the issue id and then pressing Enter, no longer works.

waltergar replied:
@vboctor why not? I think it's working properly

vboctor commented on client/js/components/IssueRelationships.tsx:
```diff
@@ -0,0 +1,144 @@
+            <button
+              onClick={() => this.handleRelationshipAdd()}
+              className='btn btn-primary btn-sm btn-white btn-round'
+            >Add</button>
```
@waltergar Replace all hard-coded strings with localized one that are now available in the html div that I added.

vboctor committed:
Emit array of relationship box buttons to HTML

Signed-off-by: Victor Boctor <victor@mantishub.net>

Commit:
hard-coded strings to localized strings

Commit:
config to react

Commit:
relationship buttons

Commit:
react check resolve bug

Commit:
removed relationship_get_summary_html

vboctor commented on config_defaults_inc.php:
```diff
@@ -1413,6 +1413,12 @@
+ * Default Bug dependent
+ * @global integer $g_default_bug_dependant
+ */
+$g_default_bug_dependant = BUG_DEPENDANT;
```
@waltergar There should be no need for us to add config options since we are trying to have parity with the existing system that didn't need such config.

vboctor commented on client/js/components/IssueRelationships.tsx:
```diff
@@ -67,22 +62,40 @@ export class IssueRelationships extends React.Component<Props, States> {
+    const { relationships } = this.state;
+
+    for (const relationship of relationships) {
+      if (relationship.type.id === this.getConfigValue('default_bug_dependant') &&
```
@waltergar We should get the config options once before the loop and have them in variables that are used within the loop.

vboctor commented:
@waltergar I assume we no longer need `bug_view_relationship_get_details()` in `bug_view_inc.php`.

vboctor commented:
@waltergar 

- Relationship box is rendered with no content after I updated and ran build.
- Resolved issues render the top and left chromes (menus/headers) but empty content.

vboctor commented:
@waltergar I assume that `bug_relationship_add.php` and `bug_relationship_delete.php` can now be deleted.

vboctor committed:
Remove dead code

- bug_view_relationship_get_details()
- bug_relationship_add.php
- bug_relationship_delete.php

Signed-off-by: Victor Boctor <victor@mantishub.net>

Commit:
fixed wrong relationships box render

vboctor commented:
@waltergar The PHP implementation had a confirmation before deleting a relationship.

```
helper_ensure_confirmed( lang_get( 'delete_relationship_sure_msg' ), lang_get( 'delete_relationship_button' ) );
```

Commit:
removed default_bug_dependant config

waltergar commented:
> The PHP implementation had a confirmation before deleting a relationship.
> 
> ```
> helper_ensure_confirmed( lang_get( 'delete_relationship_sure_msg' ), lang_get( 'delete_relationship_button' ) );
> ```

@vboctor  Do you want the exact same confirmation page or just confirm alert in javascript?

vboctor commented:
@waltergar just a JavaScript confirmation dialog.

Commit:
delete relationship confirm dialog

waltergar commented:
@vboctor , pushed the code for relationship delete confirm alert

vboctor committed:
IssueViewPageCommand: support block resolving

If an issue has related issues that it is dependant on, then block it from resolving.
This will set can_resolve and can_close flags to false and set a relationships_warning.

Signed-off-by: Victor Boctor <victor@mantishub.net>

vboctor committed:
Fix escaping of json in HTML

Signed-off-by: Victor Boctor <victor@mantishub.net>

vboctor committed:
Support auto-complete for related issue ids

Supply the auto-complete candidates from the last visited issue list.

vboctor committed:
Add internal API to get issue basic info

Commit:
input field enter key down handler

Commit:
webpack watch and clean script

vboctor committed:
Fix setting of relationship warning

vboctor committed:
Expose IssueViewPageCommand via REST API

vboctor committed:
Proper handling for relationships warning message

vboctor committed:
Add `IssueService.GetIssueBasic()`

Commit:
webpack watch

Commit:
toast message

Commit:
Merge branches 'react-relationships' and 'react-relationships' of github.com:mantishub/mantisbt into react-relationships

# Conflicts:
#	client/js/components/IssueRelationships.tsx

Commit:
removed console log

vboctor commented on client/js/components/IssueRelationships.tsx:
```diff
@@ -97,11 +103,19 @@ export class IssueRelationships extends React.Component<Props, States> {
+        try {
+          const relationships = await this.Service.RelationshipAdd(this.state.reqRelTyp, parseInt(issueId));
+          this.setState({ relationships });
+          toast.success('Success!');
```
You can use localized string `operation_successful` for this.

vboctor commented on client/js/components/IssueRelationships.tsx:
```diff
@@ -114,8 +128,12 @@ export class IssueRelationships extends React.Component<Props, States> {
       this.setState({ relationships });
-    } catch (error) {
-
+      toast.success('Success!');
```
Use localized string `operation_successful`.

vboctor commented on client/js/services/IssueService.ts:
```diff
@@ -20,10 +20,7 @@ export class IssueService {
-        throw new Error(e.response.data.message);
-      else
-        throw new Error(e);
+      throw e;
```
Why catch the exception just to throw it again? I would just remove the try/catch from both this add/delete methods for relationships.

Commit:
we don't need client/js folder. client/js -> client/

Commit:
added node_modules to gitignore

Commit:
relationships input auto-complete

Commit:
except issueId itself

Commit:
Handle arrow key control

Commit:
Arrow key input prevents cursor change

Commit:
dropdown mouse over

Commit:
working with min.js

Commit:
added watching minify script

Commit:
React plugin storybook component

Commit:
code splitting

Commit:
removed ReactSamplePlugin from Mantisbt repo

Commit:
keyboard event prevent
